### ISSUE TITLE
How to refresh shared credentials before expiry - Solution for Aws::Route53::Errors::ExpiredToken

### ISSUE THREAD

**Issue by dinesh-venkatesan** [aws/aws-sdk-ruby#2097]
### Issue description
I'm getting this error "The security token included in the request is invalid" when trying to get hosted zones list from Route53 using AWS Ruby SDK. Here is the full trace :
```
/Users/dineshvenkat/.rvm/gems/ruby-2.3.7@test/gems/aws-sdk-core-3.59.0/lib/seahorse/client/plugins/raise_response_errors.rb:15:in `call': The security token included in the request is expired (Aws::Route53::Errors::ExpiredToken)
    from /Users/dineshvenkat/.rvm/gems/ruby-2.3.7@test/gems/aws-sdk-core-3.59.0/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:20:in `call'
    from /Users/dineshvenkat/.rvm/gems/ruby-2.3.7@test/gems/aws-sdk-core-3.59.0/lib/aws-sdk-core/plugins/idempotency_token.rb:17:in `call'
    from /Users/dineshvenkat/.rvm/gems/ruby-2.3.7@test/gems/aws-sdk-core-3.59.0/lib/aws-sdk-core/plugins/param_converter.rb:24:in `call'
```

**What I have done in my code ? :** 
I have credentials file in which I will have multiple aws account credentials with their profile name. Now i will configure each and every profile credentials and gather their hosted zones.

**My ruby file is :**
```
// Reads the credentials file and gets all the profile names
File.open('/credentials_file', "r") do |file_data|
  file_data.each_line do |line|
    profile_name = line[/\[(.*?)\]/, 1]
    all_profiles << profile_name if profile_name
  end
end

// Loops each profile and gets 
all_profiles.each do |each_profile_name|
  credentials = Aws::SharedCredentials.new(profile_name: each_profile_name)
  Aws.config.update({
   region: 'us-east-1',
   credentials: credentials
  })
  route53 = Aws::Route53::Client.new

  // Gets first 100 hosted zones
  response = route53.list_hosted_zones_by_name()

  // Do something with the response
  process(response)

  // checks whether there is more hosted zones and retrieves them ( Basically pagination )
  while response.is_truncated do
    response = route53.list_hosted_zones_by_name(dns_name: response.next_dns_name, hosted_zone_id: response.next_hosted_zone_id)
    // Do something with the response
    process(response)
  end
end
```

**Problem :** 
My credentials will get expired after 1 hour of creation. So when my process takes more than 1 hour I get `Aws::Route53::Errors::ExpiredToken` error.

**My question :**
How can I overcome this ? What is the solution that you would suggest to prevent the credentials from getting expired until my process gets completed.

I have come across the concept of refreshing credentials but i have no idea on how to implement it.

Please provide a feasible solution

**Comment by cjyclaire** [comment]
Quick follow question:
I saw you are using shared config credentials providers, what type of credentials are you using in those profiles? static pairs? assume roles? web identity or others?

Our credentials themselves all implements [refresh credentials](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/refreshing_credentials.rb), but currently, we are not retrying for `Aws::Route53::Errors::ExpiredToken`, here are currently retried expired token errors https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/plugins/retry_errors.rb#L70

Then is this issue actually a request for retrying `Aws::Route53::Errors::ExpiredToken`?

**Comment by dinesh-venkatesan** [comment]
The credentials stored in the file are basically the ones that i have received from assuming roles in respective accounts.  These credentials have a validity of 60 mins. Currently I am using this file for testing purpose , however I intend to implement assume role  and getting the respective credentials as part of my ruby script instead of depending on external credential file.

Though I understand that I can have an extended time limit to the max cap of 12 hours, I wouldn't want to use it unnecessarily. I am looking for a method which could keep refreshing the credentials.

Basically ruby version of the below python boto3's implementation is what I am looking for.
https://www.pydoc.io/pypi/botocore-1.7.0/autoapi/credentials/index.html#credentials.RefreshableCredentials

**Comment by cjyclaire** [comment]
Again, Ruby credential providers (including assume role credentials) are [refreshable credentials](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/refreshing_credentials.rb), which are intended for auto-refresh when expired.

Are you looking for manual refresh(providing a `:refresh_timeout`) or current auto refresh working with Route53(retry route53 `ExpiredToken` using auto-refreshed credentials)?

**Comment by navels** [comment]
The shared credentials provider isn't refreshable, right?

**Comment by alextwoods** [comment]
`SharedCredentials` are not refreshable - they represent just static credentials loaded from a profile (NOT a set of credentials resolved from a profile using the `CredentialProviderChain` ).
